### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by YHTB**
Lack of testing and optimization

### CODE REVIEW

File: src/convert.cc [hunk 1]
```diff
@@ -0,0 +1,322 @@
...
+// 获取64位无符号整形的第一二位
+inline uint8_t GetFlag(const vint num, const uint bits = 62) {
+  uint8_t tmp = num >> bits;
+  return reinterpret_cast<uint8_t>(tmp);
+}
+inline uint8_t GetFlag(const char *const bin, const uint bits = 62) {
+  uint8_t tmp = *bin >> bits;
+  return reinterpret_cast<uint8_t>(tmp);
+}
+//获取一个字段的长度 单位是byte
```
Comment (RobertIndie) [inline src/convert.cc]: Should have a space between // and comment  [whitespace/comments]cpplint

File: src/convert.cc [hunk 2]
```diff
@@ -0,0 +1,322 @@
...
+      *frameType = FrameType::kStreamDataBlocked;
+      return GflToFsdb(gfl, reinterpret_cast<FrameStreamDataBlocked *>(frame));
+    default:
+      if (CheckFS(gfl->frame_type)) {
+        *frameType = FrameType::kStream;
+        return GflToFs(gfl, reinterpret_cast<FrameStream *>(frame));
+      }
+      return -1;
+  }
+}
```
Comment (RobertIndie) [inline src/convert.cc]: Please add a newline character at the end of the file.

File: src/convert.cc [hunk 3]
```diff
@@ -0,0 +1,322 @@
...
+//----------------------------------------------------------------------------//
+// 各种对象到二进制的转换
+enum Type { FRS = 0x04, FSDB = 0x15 };
+
+// 获取64位无符号整形的第一二位
+inline uint8_t GetFlag(const vint num, const uint bits = 62) {
+  uint8_t tmp = num >> bits;
+  return reinterpret_cast<uint8_t>(tmp);
+}
+inline uint8_t GetFlag(const char *const bin, const uint bits = 62) {
```
Comment (RobertIndie) [inline src/convert.cc]: Remove this unused function

File: src/convert.cc [hunk 4]
```diff
@@ -0,0 +1,322 @@
...
+      tmp = new char[8]{0};
+      break;
+    default:
+      return nullptr;
+  }
+  *tmp |= num;
+  return tmp;
+}
+// 合并数据,将orginData 的数据合并到location 第四个参数的单位为byte
+void MergeData(const char OrginData[], char DirData[], uint *location,
```
Comment (RobertIndie) [inline src/convert.cc]: variable names must be lowercase with underscore.
Note that DirData name is wrong, use dest_data instead.
like:
```cpp
void MergeData(const char origin_data[], char dest_data[], uint *location, uint len)

```

File: src/convert.cc [hunk 5]
```diff
@@ -0,0 +1,322 @@
...
+               uint len) {
+  for (uint i = 0; i < len; ++i) {
+    DirData[(*location)++] = OrginData[i];
+  }
+}
+struct tmpInfo {
```
Comment (RobertIndie) [inline src/convert.cc]: > Type names start with a capital letter and have a capital letter for each new word, with no underscores. -- Google C++ Style

like: TmpInfo

File: src/convert.cc [hunk 6]
```diff
@@ -0,0 +1,322 @@
...
+}
+// Frs => GFL
+void FrsToGFL(FrameResetStream *Data, GenericFrameLayout *result) {
+  // tmpInfo *data = new tmpInfo[3];
+  result->frame_type = Type::FRS;
+  //其中每一个字段的长度以及总长度
+  uint Id_len = GetLen(GetFlag(Data->stream_id));
+  uint Er_len = GetLen(GetFlag(Data->error_code));
+  uint Fi_len = GetLen(GetFlag(Data->final_size));
+  uint total_len = Id_len + Er_len + Fi_len;
```
Comment (RobertIndie) [inline src/convert.cc:181]: vint does not contain flags.

File: src/convert.cc [hunk 7]
```diff
@@ -0,0 +1,322 @@
...
+// 接受一个tmpinfo的数组，第二个参数说明需要打包几个元素，第三个参数指明数据存放到哪里
+// 为解决如何高效率的包装包
+void MergeData(const tmpInfo *const pack, uint num, char *dirData) {
+  uint location = 0;
+  for (uint i = 0; i < num; ++i) {
+    MergeData((*(pack + i)).data, dirData, &location, (*(pack + i)).len);
+  }
+}
+// Frs => GFL
+void FrsToGFL(FrameResetStream *Data, GenericFrameLayout *result) {
```
Comment (RobertIndie) [inline src/convert.cc]: Abbreviated name must be all capital.
like: FRSToGFL

**Review by RobertIndie — CHANGES_REQUESTED** [review]
Please fix and test convert.cc

**Review by RobertIndie — CHANGES_REQUESTED** [review]
Compile Error. Please fix it.

File: src/convert.cc [hunk 8]
```diff
@@ -0,0 +1,452 @@
...
+  } else if (Rang<uint32_t>(*num)) {
+    return TypeSize<uint32_t>();
+  } else if (Rang<uint64_t>(*num)) {
+    return TypeSize<uint64_t>();
+  }
+  return 0;
+}
+// 获得标志位 这些标志位放在小端 使用时候需要根据需求来向左移相应的位数
+// 若没有匹配到相应的字节数那么返回0xff
+uint8_t GetFlag(uint DataLen) {
```
Comment (RobertIndie) [inline src/convert.cc]: use vint, not uint

File: src/convert.cc [hunk 9]
```diff
@@ -0,0 +1,452 @@
...
+    return TypeSize<uint8_t>();
+  } else if (Rang<uint16_t>(*num)) {
+    return TypeSize<uint16_t>();
+  } else if (Rang<uint32_t>(*num)) {
+    return TypeSize<uint32_t>();
+  } else if (Rang<uint64_t>(*num)) {
+    return TypeSize<uint64_t>();
+  }
+  return 0;
+}
```
Comment (RobertIndie) [inline src/convert.cc]: Not in the correct range.

2Bit | Length | Usable Bits | Range
-- | -- | -- | --
00 | 1 | 6 | 0-63
01 | 2 | 14 | 0-16383
10 | 4 | 30 | 0-1073741823
11 | 8 | 62 | 0-4611686018427387903

File: src/convert.cc [hunk 10]
```diff
@@ -0,0 +1,452 @@
...
+void SetFlag(vint *data) {
+  uint MinAir = GetMinLen(data);
+  vint tmpFlag = GetFlag(MinAir);
+  if (tmpFlag == 0xff) {
+    std::cout << "DataFiled" << std::endl;
+    return;  // 终止程序
+  }
+  tmpFlag <<= 62;  // 将标志位移动到应该设置的位置
+  *data |= tmpFlag;
+}
```
Comment (RobertIndie) [inline src/convert.cc:60]: Bug

**Review by RobertIndie — CHANGES_REQUESTED** [review]

File: src/convert.cc (hunk 8 above)
Comment (YHTB) [reply src/convert.cc]: Done

**Pull request merged**